### PR TITLE
Add Watchlist Filters support

### PR DESCRIPTION
**Pull request by tonisevener**
This is the (hopefully all-encompassing) data portion needed to support Watchlist Filters. Since this introduces our first persistence logic in the data framework, I decided to rename things a bit so that the purpose of each of our classes felt a little more clear. I was falling into the trap of naming everything a service without this adjustment. These are the new terms I'm going for:

**Service** - Services indicate classes that trigger some sort of call to the network. I do not intend for a service to be feature-specific, but we _could_ have multiple services for each backend system. As of now we only have a MediaWiki Service for calling MediaWiki APIs, which the client app implements using a Fetcher subclass. Eventually we hope to move it into the data framework. We could eventually make additional services for calling to the analytics backend, or RESTBase backend, etc.

**Store** - Stores indicate classes that persist stuff to device. Like services, they will not be feature-specific, but we _could_ have multiple stores that go to different local systems like UserDefaults, SharedContainerCache, Core Data, etc. Similar to how the client app actually implements the MediaWiki service with a fetcher, the client app will implement the cache store using its existing SharedContainerCache. I went ahead worked a bit on the caching portion of the offline mode [task](https://phabricator.wikimedia.org/T335817) in this PR to be sure that we could lean on the same setup for both persisting filter settings and watchlist items.

**DataController** - These will be classes that are feature-specific and internally ask the appropriate services and stores for data. These data controllers will be what Components call to get their data. Totally open to a different name here, I just didn't want it to end in Store or Service.

Please first merge https://github.com/wikimedia/wikipedia-ios-components/pull/19 to avoid compiler issues once this gets merged.

**Test Steps**
Confirm unit tests pass. The rest will be tested in upcoming PRs in the Components and client repos.

### CODE REVIEW

File: Sources/WKData/Environment/WKDataEnvironment.swift [hunk 1]
```diff
@@ -13,5 +13,7 @@ public final class WKDataEnvironment: ObservableObject {
 	public static let current = WKDataEnvironment()
 
     @Published public var appData = WKAppData(appLanguages: [])
-    @Published public var mediaWikiNetworkService: WKNetworkService?
+    public var mediaWikiService: WKService?
+    public var userDefaultsStore: WKKeyValueStore? = WKUserDefaultsStore()
```
Comment (staykids) [inline Sources/WKData/Environment/WKDataEnvironment.swift]: Other than for unit test purposes, any reason for this not to be immutable? Is there a reason we'd want the client app to inject this with something else?

File: Sources/WKData/Models/WKProject.swift [hunk 2]
```diff
@@ -1,6 +1,21 @@
...
+        case .wikipedia(let language):
+            if let languageVariantCode = language.languageVariantCode {
+                return "wikipedia-" + language.languageCode + "-" + languageVariantCode
+            } else {
+                return "wikipedia-" + language.languageCode
+            }
+        }
+    }
+    
     public static func == (lhs: WKProject, rhs: WKProject) -> Bool {
```
Comment (staykids) [inline Sources/WKData/Models/WKProject.swift:19]: Am I missing a reason we can't simplify this conformance to just `return lhs.id == rhs.id`?

File: Tests/WKDataTests/WKUserDefaultsStoreTests.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
+//
+//  WKUserDefaultsStoreTests.swift
+//  
+//
+//  Created by Toni Sevener on 8/8/23.
+//
+
```
Comment (staykids) [inline Tests/WKDataTests/WKUserDefaultsStoreTests.swift]: Minor: file header snuck in here.
```suggestion
```

**Review by staykids — COMMENTED** [review]
Nice – minor notes below and a merge conflict to address.

File: Sources/WKData/Environment/WKDataEnvironment.swift (hunk 1 above)
Comment (tonisevener) [reply Sources/WKData/Environment/WKDataEnvironment.swift]: Nope, no reason other than for unit tests. I added an `internal(set)` to this, so client apps are no longer allowed to change it.  Let me know if you have other ideas on locking it down.

File: Sources/WKData/Models/WKProject.swift (hunk 2 above)
Comment (tonisevener) [reply Sources/WKData/Models/WKProject.swift:19]: Nope, this is updated! I think the Equatable conformance came before Identifiable, and I didn't go back and simplify. Good catch.

File: Tests/WKDataTests/WKUserDefaultsStoreTests.swift (hunk 3 above)
Comment (tonisevener) [reply Tests/WKDataTests/WKUserDefaultsStoreTests.swift]: Fixed!

**Review by staykids — APPROVED** [review]

**Pull request merged**